File: app/src/main/java/com/stho/software/nyota/utilities/Moment.java
package com.stho.software.nyota.utilities;

import java.util.Calendar;
import java.util.TimeZone;

/**
 * Created by shoedtke on 16.09.2016.
 */
public class Moment {

    private final UTC utc;
    private final City city;

    public final double LST; // in hours [0 ; 24]
    public final double d;

    // only work with UTC time, timeZone = GMT
    // when printing time for a moment, it means to print time for a timeZone, so use calendar there when displaying the time. Only then.

    public Moment(final City city, final UTC utc) {
        this.utc = utc; // assuming utc.getTimeZone() is always GMT
        this.city = city;
        this.LST = utc.getLST(city.getLongitude());
        this.d = utc.getDayNumber();
    }

    public Moment(final City city, final long timeInMillis) {
        this.utc = new UTC(timeInMillis);
        this.city = city;
        this.LST = utc.getLST(city.getLongitude());
        this.d = utc.getDayNumber();
    }


    @Override
    public String toString() {
        return Formatter.toString(utc.getTime(), city.getTimeZone(), Formatter.TimeFormat.DATETIME_SEC_TIMEZONE);
    }

    public Moment forNow() {
        return new Moment(city, UTC.forNow());
    }

    /**
     * Keep current location and timezone, but set a new time
     * @param timeInMillis UTC in milliseconds
     * @return a new moment instance
     */
    Moment forUTC(final long timeInMillis) {
        return new Moment(city, new UTC(timeInMillis));
    }

    /**
     * Keep current location and timezone, but set a new time
     * @param utc UTC
     * @return a new moment instance
     */
    public Moment forUTC(final UTC utc) {
        return new Moment(city, utc);
    }

    /**
     * Keep current location and timezone, but set a new time by adding milliseconds
     * @param millis
     * @return
     */
    public Moment addMillis(final long millis) {
        return new Moment(city, utc.addMillis(millis));
    }

    /**
     * Keep current location and timezone, but set a new time by adding hours
     * @param hours any time difference in hours
     * @return a new moment instance
     */
    public Moment addHours(final double hours) {
        return new Moment(city, utc.addHours(hours));
    }

    /**
     * Keep current location and timezone, but set a new time by adding months
     * @param months
     * @return
     */
    public Moment addMonths(final int months) {
        Calendar calendar = getLocalTime();
        calendar.add(Calendar.MONTH, months);
        return forUTC(calendar.getTimeInMillis());
    }

    /**
     * Keep current location and timezone, but set a new time by adding years
     * @param years
     * @return
     */
    public Moment addYears(final int years) {
        Calendar calendar = getLocalTime();
        calendar.add(Calendar.YEAR, years);
        return forUTC(calendar.getTimeInMillis());
    }

    public static Moment forUTC(final City city, final UTC utc) {
        return new Moment(city, utc);
    }

    public static Moment forNow(final City city) {
        return new Moment(city, UTC.forNow());
    }

    public Location getLocation() {
        return city.getLocation();
    }

    public TimeZone getTimeZone() {
        return city.getTimeZone();
    }

    public UTC getUTC() {
        return utc;
    }

    public City getCity() {
        return city;
    }

    public Calendar getLocalTime() {
        final Calendar calendar = Calendar.getInstance(city.getTimeZone());
        calendar.setTimeInMillis(utc.getTimeInMillis());
        return calendar;
    }
}

File: app/src/main/java/com/stho/nyota/sky/utilities/Moment.kt
package com.stho.nyota.sky.utilities

import com.stho.nyota.Interval
import java.util.*

/**
 * Created by shoedtke on 16.09.2016.
 */
class Moment(val city: City, override val utc: UTC) : IMoment {

    override val lst: Double = utc.getLST(city.longitude) // in hours [0 ; 24]
    override val d: Double = utc.dayNumber

    override val location: Location
        get() = city.location

    override val timeZone: TimeZone
        get() = city.timeZone

    val version: Long = System.nanoTime()

    val localTime: Calendar
        get() {
            val calendar = Calendar.getInstance(city.timeZone)
            calendar.timeInMillis = utc.timeInMillis
            return calendar
        }

    override fun toString(): String =
        Formatter.toString(utc.time, city.timeZone, Formatter.TimeFormat.DATETIME_SEC_TIMEZONE)

    fun forNow(): Moment =
        Moment(city, UTC.forNow())

    override fun forUTC(timeInMillis: Long): IMoment =
        Moment(city, UTC.forTimeInMillis(timeInMillis))

    override fun forUTC(newUtc: UTC): Moment =
        Moment(city, newUtc)

    fun forThisDate(year: Int, month: Int, dayOfMonth: Int): Moment {
        val calendar = localTime
        calendar.set(Calendar.YEAR, year)
        calendar.set(Calendar.MONTH, month)
        calendar.set(Calendar.DAY_OF_MONTH, dayOfMonth)
        return Moment(city, UTC.forCalendar(calendar))
    }

    fun forThisTime(hourOfDay: Int, minute: Int): Moment {
        val calendar = localTime
        calendar.set(Calendar.HOUR_OF_DAY, hourOfDay)
        calendar.set(Calendar.MINUTE, minute)
        calendar.set(Calendar.SECOND, 0)
        return Moment(city, UTC.forCalendar(calendar))
    }

    /**
     * Keep current location and timezone, but set a new time by adding milliseconds
     * @param millis
     * @return
     */
    internal fun addMillis(millis: Long): Moment {
        return Moment(city, utc.addMillis(millis))
    }

    /**
     * Keep current location and timezone, but set a new time by adding hours
     * @param hours any time difference in hours
     * @return a new moment instance
     */
    internal fun addHours(hours: Double): Moment {
        return Moment(city, utc.addHours(hours))
    }

    /**
     * Keep current location and timezone, but set a new time by adding months
     * @param months
     * @return
     */
    internal fun addMonths(months: Int): Moment {
        val calendar = localTime
        calendar.add(Calendar.MONTH, months)
        return Moment(city, UTC.forCalendar(calendar))
    }

    /**
     * Keep current location and timezone, but set a new time by adding years
     * @param years
     * @return
     */
    internal fun addYears(years: Int): Moment {
        val calendar = localTime
        calendar.add(Calendar.YEAR, years)
        return Moment(city, UTC.forCalendar(calendar))
    }

    internal fun next(interval: Interval): Moment =
        when (interval) {
            Interval.MINUTE -> addMillis(60000)
            Interval.HOUR -> addHours(1.0)
            Interval.DAY -> addHours(24.0)
            Interval.MONTH -> addMonths(1)
            Interval.YEAR -> addYears(1)
        }


    internal fun previous(interval: Interval): Moment =
        when (interval) {
            Interval.MINUTE -> addMillis(-60000)
            Interval.HOUR -> addHours(-1.0)
            Interval.DAY -> addHours(-24.0)
            Interval.MONTH -> addMonths(-1)
            Interval.YEAR -> addYears(-1)
        }


    companion object {

        fun forUTC(city: City, utc: UTC): Moment =
            Moment(city, utc)

        fun forUTC(city: City, timeInMillis: Long): Moment =
            Moment(city, UTC.forTimeInMillis(timeInMillis))

        fun forNow(city: City): Moment =
            Moment(city, UTC.forNow())
    }
}